### PR TITLE
feat: add rate limiter

### PR DESCRIPTION
**Pull request by raphael0202**
Add a rate-limiter based on Redis.

### CODE REVIEW

File: cgi/display.pl [hunk 1]
```diff
@@ -117,6 +117,13 @@
 	return Apache2::Const::OK;
 }
 
+if ($request_ref->{rate_limiter_blocking} eq 1) {
```
Comment (stephanegigandet) [inline cgi/display.pl:120]:
```suggestion
if ($request_ref->{rate_limiter_blocking} ) {
```

eq is for string comparison

File: lib/ProductOpener/Redis.pm [hunk 2]
```diff
@@ -143,4 +145,130 @@ sub push_to_redis_stream ($user_id, $product_ref, $action, $comment, $diffs) {
...
+		# we were disconnected, try again
+		$log->info("Trying to reconnect to Redis");
+		init_redis();
+	}
+	my $resp;
+	if (defined $redis_client) {
+		$log->debug("Getting rate-limit remaining requests", {ip => $ip, api_action => $api_action})
+			if $log->is_debug();
+		my $current_minute = int(time() / 60);
+		eval {$resp = $redis_client->get("po-rate-limit:$ip:$api_action:$current_minute");};
```
Comment (alexgarel) [inline lib/ProductOpener/Redis.pm:186]: I would put key computation in a function, just to be sure we don't mess in a subtle way with it at some point (like in a refactor)...

**Review by alexgarel — COMMENTED** [review]
Cool, I think it's almost ready for a release.

But:
1. we must add white listing of some ips (for our own server like robotoff or facets knowledge panels, and maybe other servers). [NET::IP::Match](https://metacpan.org/pod/Net::IP::Match) could help, but a simple hashmap with keys is also ok I think :-)
2. @stephanegigandet do you want to handle merge conflict (as those are mostly due to your refactors, I think)

File: lib/ProductOpener/Config_off.pm [hunk 3]
```diff
@@ -1549,4 +1555,11 @@ $options{sample_product_code} = "093270067481501";    # A good product for you -
 #$options{sample_product_code_language_de} = "20884680"; # Waffeln Sondey
 #$options{sample_product_code_country_at_language_de} = "5411188119098"; # Natur miss kokosnuss Alpro
 
+## Rate limiting ##
+
+# Number of requests per minutes for the search API
+$options{rate_limit_search} = 2;
+# Number of requests per minutes for the product API
+$options{rate_limit_product} = 10;
```
[the remark this answers is not in the document]
Comment (raphael0202) [reply lib/ProductOpener/Config_off.pm:1563]: Indeed, it was for test purposes actually. The rate limits are in the docs : https://openfoodfacts.github.io/openfoodfacts-server/api/#rate-limits
100 req/min for product queries
10 req/min for search queries

File: cgi/search.pl [hunk 4]
```diff
@@ -65,6 +66,17 @@
 
 my $request_ref = ProductOpener::Display::init_request();
 $request_ref->{search} = 1;
+# api_action is required for `check_and_update_rate_limits`
+$request_ref->{api_action} = 'search';
+
+check_and_update_rate_limits($request_ref);
+
+if ($request_ref->{rate_limiter_blocking} eq 1) {
```
Comment (raphael0202) [inline cgi/search.pl:74]:
```suggestion
if ($request_ref->{rate_limiter_blocking}) {
```

File: cgi/display.pl [hunk 5]
```diff
@@ -117,6 +117,13 @@
 	return Apache2::Const::OK;
 }
 
+if ($request_ref->{rate_limiter_blocking} ) {
```
Comment (raphael0202) [inline cgi/display.pl:120]:
```suggestion
if ($request_ref->{rate_limiter_blocking}) {
``

**Comment by teolemon** [comment]
@raphael0202 merge conflicts

**Comment by stephanegigandet** [comment]
I'll fix the conflicts

**Review by stephanegigandet — APPROVED** [review]
Thanks @raphael0202 . I tested locally, it works fine. Limits need to be increased as @alexgarel said.

**Comment by raphael0202** [comment]
I've increased the limits, it looks good to ship (in non-blocking mode for now, so that we can at least identify how frequent API abusers are).

**Comment by teolemon** [comment]
@raphael0202 @stephanegigandet 
The following jobs failed:
+--------------------------------------+----------------------+
| Job ID                               | Test File            |
+--------------------------------------+----------------------+
| B02C803E-0604-11EF-9E5F-A7BACFC7AEAC | tests/unit/routing.t |
+--------------------------------------+----------------------+

                                    Yath Result Summary
-------------------------------------------------------------------------------------------
     Fail Count: 1
     File Count: 66
Assertion Count: 5997
      Wall Time: 534.14 seconds
       CPU Time: 2031.37 seconds (usr: 4.80s | sys: 0.44s | cusr: 1801.00s | csys: 225.13s)
      CPU Usage: 380%
    -->  Result: FAILED  <--

**Pull request merged**